### PR TITLE
w for  custom charset

### PR DESCRIPTION
**Pull request by gogo2464**
**Checklist**

- [ ] Closing issues: #issue
- [ ] Mark this if you consider it ready to merge
- [ ] I've added tests (optional)
- [ ] I wrote some lines in the [radare2book](https://github.com/radareorg/radare2book)

**Description**

### CODE REVIEW

File: libr/core/cconfig.c [hunk 1]
```diff
@@ -1233,6 +1233,20 @@ static bool cb_cfgcharset(void *user, void *data) {
 		char *syscs = r_str_newf ("%s%s.sdb", cs, cf);
 		if (r_file_exists (syscs)) {
 			rc = r_charset_open (core->print->charset, syscs);
+
+			SdbListIter *iter;
+			SdbKv *kv;
+			SdbList *sdbls = sdb_foreach_list (core->print->charset->db, true);
+			Sdb *db = sdb_new0();
```
Comment (trufae) [inline libr/core/cconfig.c:1240]:
```suggestion
			Sdb *db = sdb_new0 ();
```

File: libr/core/cconfig.c [hunk 2]
```diff
@@ -1233,6 +1233,20 @@ static bool cb_cfgcharset(void *user, void *data) {
 		char *syscs = r_str_newf ("%s%s.sdb", cs, cf);
 		if (r_file_exists (syscs)) {
 			rc = r_charset_open (core->print->charset, syscs);
+
```
Comment (trufae) [inline libr/core/cconfig.c:1236]: shouldnt this code be moved into r_charset_open() instead ? i dont see why it should be done in the config callback

**Review by trufae — CHANGES_REQUESTED** [review]
a

**Comment by gogo2464** [comment]
@trufae the rebase on the latest version of r2 does not work. I am waiting for a pull request to solve a compilation error from esil:


```
=hidden -o anal.o anal.c
anal.c:17:2: error: ‘R_ESIL_STATIC_PLUGINS’ undeclared here (not in a function); did you mean ‘R_ANAL_STATIC_PLUGINS’?
  R_ESIL_STATIC_PLUGINS
  ^~~~~
  R_ANAL_STATIC_PLUGINS
```

**Comment by gogo2464** [comment]
The latest commit is just to pass the travis tests and see. Nothing new to show. I am going to work.

File: libr/core/cconfig.c (hunk 2 above)
Comment (trufae) [reply libr/core/cconfig.c:1236]: move the dbg_char_to_hex generation code inside the r_charset_open() function

File: libr/core/cmd_write.c [hunk 3]
```diff
@@ -2017,8 +2017,29 @@ static int cmd_write(void *data, const char *input) {
 		wc_handler_old (core, input + 1);
 		break;
 	case ' ': // "w"
-		w_handler_old (core, input + 1);
+	{
+		size_t len = core->blocksize;
+		const char *current_charset = r_config_get (core->config, "cfg.charset");
+		if (R_STR_ISEMPTY (current_charset)) {
+			w_handler_old (core, input + 1);
```
Comment (trufae) [inline libr/core/cmd_write.c:2024]: imho the logic of text encoding should be done in a diferent place. see the w_handle_old code and make the encoding there, not in every caller

**Review by trufae — CHANGES_REQUESTED** [review]

File: libr/core/cmd_write.c (hunk 3 above)
Comment (gogo2464) [reply libr/core/cmd_write.c:2024]: @trufae I did same for the command ps. I might change ps too.

Comment (trufae) [reply libr/core/cmd_write.c:2024]: yes

File: libr/core/cconfig.c (hunk 2 above)
Comment (trufae) [reply libr/core/cconfig.c:1236]: see this  ^

**Comment by gogo2464** [comment]
r_charset_encode_str writes tons of ???. I can not use it for w. Do I write another function?

**Comment by trufae** [comment]
So what would u expect to do in case the char cannot be converted? imho just having a single function and adding a parameter to specify what to do in that case would be enough.

so the options for this argument would be: 

* set ?
* skip char
* add a space or null byte
* keep the original value

**Comment by gogo2464** [comment]
@trufae We might want to use a function that only print recognized chars if they are recognized, no char if no char recognized and stops when no char is recognized anymore. Exactly like ps with no e cfg.charset=pokered set. 

My code used to do that. It has been changed and replaced by ????. So I think it is not what you need. So I purpose to keep the ???? for ps but make a function that works only like native ps and use it in another function called when ps with e cfg.charset is set.

**Comment by trufae** [comment]
We discussed this ?? Thing and its a parameter to select what do when the char cant be transcoded. And maybe good to have an user option for this but expected behaviour should be the default.

**Comment by gogo2464** [comment]
@trufae we did not talk on this. I just let it like that in my first PR 16621. This PR was incomplete and you finished it. Your changes were great! I also realize that it has changed this behavior. This is all.

**Comment by trufae** [comment]
So whats left to finish this pr? I would like to merge it when ready and test it more to give more feedback and ideas

**Comment by gogo2464** [comment]
@trufae the program simply does not work yet. I have to fix it.

File: libr/core/cmd_print.c [hunk 4]
```diff
@@ -5886,7 +5886,7 @@ l = use_blocksize;
 							ut8 *data = malloc (len);
 							if (data) {
 								r_io_read_at (core->io, core->offset, data, len);
-								r_charset_encode_str (core->print->charset, out, out_len, data, len);
+								r_charset_encode_str (core->print->charset->db, out, out_len, data, len, false);
```
Comment (trufae) [inline libr/core/cmd_print.c:5889]: nope
```suggestion
								r_charset_encode_str (core->print->charset, out, out_len, data, len, false);
```

File: libr/core/cmd_print.c [hunk 5]
```diff
@@ -5797,7 +5797,7 @@ l = use_blocksize;
 						ut8 *data = malloc (len);
 						if (data) {
 							r_io_read_at (core->io, core->offset, data, len);
-							r_charset_encode_str (core->print->charset, out, out_len, data, len);
+							r_charset_encode_str (core->print->charset->db, out, out_len, data, len, false);
```
Comment (trufae) [inline libr/core/cmd_print.c:5800]:
```suggestion
							r_charset_encode_str (core->print->charset, out, out_len, data, len, false);
```

File: libr/core/cconfig.c [hunk 6]
```diff
@@ -1233,8 +1234,40 @@ static bool cb_cfgcharset(void *user, void *data) {
...
 		} else {
 			rc = r_charset_open (core->print->charset, cf);
+
+			sdb_reset (core->print->charset->db_char_to_hex);
+			sdb_open (core->print->charset->db_char_to_hex, cs);
+			SdbListIter *iter;
+			SdbKv *kv;
+			SdbList *sdbls = sdb_foreach_list (core->print->charset->db, true);
+			Sdb *db = sdb_new0();
```
Comment (trufae) [inline libr/core/cconfig.c:1263]:
```suggestion
			Sdb *db = sdb_new0 ();
```

File: libr/core/cconfig.c [hunk 7]
```diff
@@ -1233,8 +1234,40 @@ static bool cb_cfgcharset(void *user, void *data) {
...
+			ls_foreach (sdbls, iter, kv) {
+				const char *new_key = kv->base.value;
+				const char *new_value = kv->base.key;
+				sdb_add (db, new_value, new_key, 0);
+			}
+			ls_free (sdbls);
+			core->print->charset->db_char_to_hex = db;
```
Comment (trufae) [inline libr/core/cconfig.c:1270]: why are you calling reset+open and then replacing the pointer? this is leaking . you should sdb_free() before assigning that

File: libr/core/cconfig.c [hunk 8]
```diff
@@ -1233,8 +1234,40 @@ static bool cb_cfgcharset(void *user, void *data) {
...
+				sdb_add (db, new_key, new_value, 0);
+			}
+			ls_free (sdbls);
+			core->print->charset->db_char_to_hex = db;
+
+
 		} else {
 			rc = r_charset_open (core->print->charset, cf);
+
+			sdb_reset (core->print->charset->db_char_to_hex);
```
Comment (trufae) [inline libr/core/cconfig.c:1258]: this is leaking hard:
* if you reset theres no need to open to initialize it
* just remove the sdb_reset and sbd_open calls

File: libr/core/cconfig.c [hunk 9]
```diff
@@ -1233,8 +1234,40 @@ static bool cb_cfgcharset(void *user, void *data) {
...
+			sdb_open (core->print->charset->db_char_to_hex, cs);
+			SdbListIter *iter;
+			SdbKv *kv;
+			SdbList *sdbls = sdb_foreach_list (core->print->charset->db, true);
```
Comment (trufae) [inline libr/core/cconfig.c:1262]: _list methods are slow

File: libr/core/cmd_write.c [hunk 10]
```diff
@@ -2025,8 +2025,29 @@ static int cmd_write(void *data, const char *input) {
 		wA_handler_old (core, input + 1);
 		break;
 	case ' ': // "w"
-		w_handler_old (core, input + 1);
+	{
+		size_t len = core->blocksize;
+		const char *current_charset = r_config_get (core->config, "cfg.charset");
```
Comment (trufae) [inline libr/core/cmd_write.c:2030]:
```suggestion
		const char *cur = r_config_get (core->config, "cfg.charset");
```

Comment (trufae) [reply libr/core/cmd_write.c:2030]: use cur or curcs, seems like better name to me

File: libr/core/cmd_write.c [hunk 11]
```diff
@@ -2025,8 +2025,29 @@ static int cmd_write(void *data, const char *input) {
...
+		if (R_STR_ISEMPTY (current_charset)) {
+			w_handler_old (core, input + 1);
+		} else {
+			if (len > 0) {
+				size_t in_len = strlen (input + 1);
+				ut8 *out = malloc (len); //suppose in len = out len TODO: change it
```
Comment (trufae) [inline libr/core/cmd_write.c:2036]: imho this logic shouldbe implemented inside w_handler_old() just when trying to write the string

File: libr/include/r_util/r_str.h [hunk 12]
```diff
@@ -42,7 +43,7 @@ R_API RCharset *r_charset_new(void);
 R_API void r_charset_free(RCharset *charset);
 R_API RCharsetRune *r_charset_rune_new(const ut8 *ch, const ut8 *hx);
 R_API void r_charset_rune_free(RCharsetRune *rcr);
-R_API size_t r_charset_encode_str(RCharset *r_char, ut8 *out, size_t out_len, const ut8 *in, size_t len_input);
+R_API size_t r_charset_encode_str(Sdb *sdb, ut8 *out, size_t out_len, const ut8 *in, size_t in_len, bool opt_stop_at_error);
```
Comment (trufae) [inline libr/include/r_util/r_str.h:46]: revert

File: libr/util/charset.c [hunk 13]
```diff
@@ -76,15 +76,25 @@ R_API RCharsetRune *search_from_hex(RCharsetRune *r, const ut8 *hx) {
 }
 
 // assumes out is as big as in_len
-R_API size_t r_charset_encode_str(RCharset *rc, ut8 *out, size_t out_len, const ut8 *in, size_t in_len) {
+R_API size_t r_charset_encode_str(Sdb *sdb, ut8 *out, size_t out_len, const ut8 *in, size_t in_len, bool opt_stop_at_error) {
```
Comment (trufae) [inline libr/util/charset.c:79]: revert

**Review by trufae — CHANGES_REQUESTED** [review]

File: libr/core/cmd_print.c (hunk 4 above)
Comment (gogo2464) [reply libr/core/cmd_print.c:5889]: @trufae I planned to use r_charset_decode for both the ps (charset->db) and to w (charset->db_char_to_hex).

**Comment by gogo2464** [comment]
w just works but I see some improvements to do:

-exchange names of r_charset_decode_str and r_charset_encode_str
-add a new function to refactor the changes on cb_cfgcharset

File: libr/core/cconfig.c [hunk 14]
```diff
@@ -1220,6 +1220,7 @@ static void list_available_plugins(const char *path) {
 static bool cb_cfgcharset(void *user, void *data) {
 	RCore *core = (RCore*) user;
 	RConfigNode *node = (RConfigNode*) data;
```
Comment (trufae) [inline libr/core/cconfig.c:1222]: Unused?

File: libr/core/cconfig.c [hunk 15]
```diff
@@ -1231,10 +1232,34 @@ static bool cb_cfgcharset(void *user, void *data) {
...
+			ls_foreach (sdbls, iter, kv) {
+				const char *new_key = kv->base.value;
+				const char *new_value = kv->base.key;
+				//snprintf (k, sizeof (k), "0x%02x", (unsigned int) new_key);
+				sdb_add (core->print->charset->db_char_to_hex, new_key, new_value, 0);
+			}
+			//const char *to_debug = sdb_const_get (core->print->charset->db_char_to_hex, "A", 0);
+			ls_free (sdbls);
 		} else {
-			rc = r_charset_open (core->print->charset, cf);
```
Comment (trufae) [inline libr/core/cconfig.c:1237]: The else block is the same as the if block the only thing that changes is the open. Make it simpler

File: libr/core/cconfig.c [hunk 16]
```diff
@@ -1231,10 +1231,34 @@ static bool cb_cfgcharset(void *user, void *data) {
 		list_available_plugins (cs);
 	} else {
 		char *syscs = r_str_newf ("%s%s.sdb", cs, cf);
+		core->print->charset->db_char_to_hex = sdb_new0 ();
```
Comment (trufae) [inline libr/core/cconfig.c:1234]: This line should be inside the charset_open function

File: libr/core/cconfig.c [hunk 17]
```diff
@@ -1231,10 +1231,34 @@ static bool cb_cfgcharset(void *user, void *data) {
...
+				const char *new_value = kv->base.key;
+				//snprintf (k, sizeof (k), "0x%02x", (unsigned int) new_key);
+				sdb_add (core->print->charset->db_char_to_hex, new_key, new_value, 0);
+			}
+			//const char *to_debug = sdb_const_get (core->print->charset->db_char_to_hex, "A", 0);
+			ls_free (sdbls);
 		} else {
-			rc = r_charset_open (core->print->charset, cf);
+			rc = r_charset_open (core->print->charset, syscs);
+			SdbListIter *iter;
```
Comment (trufae) [inline libr/core/cconfig.c:1251]: This code that fills the second sdb mist be inside the charset_open function

File: libr/core/cconfig.c [hunk 18]
```diff
@@ -1231,10 +1231,34 @@ static bool cb_cfgcharset(void *user, void *data) {
...
+			SdbKv *kv;
+			SdbList *sdbls = sdb_foreach_list (core->print->charset->db, true);
+
+			ls_foreach (sdbls, iter, kv) {
```
Comment (trufae) [inline libr/core/cconfig.c:1255]: Same

File: libr/core/cmd_print.c [hunk 19]
```diff
@@ -5793,7 +5793,7 @@ l = use_blocksize;
 						ut8 *data = malloc (len);
 						if (data) {
 							r_io_read_at (core->io, core->offset, data, len);
-							r_charset_encode_str (core->print->charset, out, out_len, data, len);
+							r_charset_encode_str (core->print->charset->db, out, out_len, data, len);
```
Comment (trufae) [inline libr/core/cmd_print.c:5796]: Pass the charset struct. Not the db

File: libr/core/cmd_write.c [hunk 20]
```diff
@@ -2025,8 +2025,24 @@ static int cmd_write(void *data, const char *input) {
 		wA_handler_old (core, input + 1);
 		break;
 	case ' ': // "w"
-		w_handler_old (core, input + 1);
+	{
+		size_t len = core->blocksize;
+		const char *curcs= r_config_get (core->config, "cfg.charset");
```
Comment (trufae) [inline libr/core/cmd_write.c:2030]: Space before =

File: libr/core/cmd_write.c [hunk 21]
```diff
@@ -2025,8 +2025,24 @@ static int cmd_write(void *data, const char *input) {
...
+		if (R_STR_ISEMPTY (curcs)) {
+			w_handler_old (core, input + 1);
+		} else {
+			if (len > 0) {
+				size_t in_len = strlen (input + 1);
+				ut8 *out = malloc (in_len); //suppose in len = out len TODO: change it
+				if (out) {
+					r_charset_decode_str (core->print->charset->db_char_to_hex, out, in_len, (const unsigned char *) input + 1, in_len);
```
Comment (trufae) [inline libr/core/cmd_write.c:2038]: Pass the charset struct not the db

**Review by trufae — CHANGES_REQUESTED** [review]

**Comment by gogo2464** [comment]
All tests have passed. It should be ready to merge.

**Review by trufae — APPROVED** [review]
Looks good! i'll add a few comments for future PRs, but the feature and implementation looks great!

**Pull request merged**